### ISSUE TITLE
Allow notifications (info/warn/error) to be cancelled/hidden

### ISSUE THREAD

**Issue by DanTup** [Microsoft/vscode#46551]
Sometime I pop up notifications based on the state of the project, like this one to offer to restore packages:

<img width="492" alt="screen shot 2018-03-26 at 14 18 11" src="https://user-images.githubusercontent.com/1078012/37908627-b098cf3a-3100-11e8-8286-43f18299d8fd.png">

Sometimes the user performs another action that makes this redundant (for example when you launch a Flutter app with `F5` it automatically restores packages as it launches). It would be nice if there was a way to cancel the notification so it disappears on its own rather than hanging around when it makes no more sense.

**Comment by bpasero** [comment]
@jrieken not sure if this has been asked already. I think it would be nice if extensions could hide a notification but I am not sure how to do this with the current API that returns a `Thenable`.

**Comment by DanTup** [comment]
Is there some sort of cancellation token we could pass into MessageOptions maybe?

**Comment by DanTup** [comment]
Or maybe something like `window.withProgress` which takes a promise that can be resolved/rejected?

**Comment by bpasero** [comment]
@DanTup a `CancellationToken` is typically the other way around to inform you that the user canceled the operation. But yeah, if you could pass over a cancellation token source, the cancellation could be wired in from there.

**Comment by jrieken** [comment]
Not a fan. A notification is a strong distractor and once extensions show them they grab the attention of the user. Giving them the opportunity to hide them isn't something I'd like us to support because the worst would be notifications that come and go and keep distracting me.

**Comment by jrieken** [comment]
/duplicate of https://github.com/Microsoft/vscode/issues/44122

**Comment by DanTup** [comment]
@jrieken 

> Not a fan. A notification is a strong distractor and once extensions show them they grab the attention of the user. Giving them the opportunity to hide them isn't something I'd like us to support because the worst would be notifications that come and go and keep distracting me.

I understand, but there are many places where we end up with a worse user experience.

For example, I was just implementing some refactors, like `Extract Method`. If the user tries to name their method with invalid characters, I show an error. The modal errors look terrible so I use this notification. If the user reads the notification in the corner of the screen then decides to just re-invoke the the refactor and give it a good name, they have a stale error message hanging around which is confusing.

I know there are shortcuts to get to the notifications, but many users won't know them, they'll just get annoyed that these error messages hang around after they've already dealt with them.

**Comment by DanTup** [comment]
I noticed in a fix for something else there's a hideAfterTimeout call in the notifications. I've never seen a notification hide on its own (except for the bug I raised). Is there something specific I need to do to make sure they're purged in this code? Does it only happen if there are no buttons (I can't see anything obvious in [the code](https://github.com/Microsoft/vscode/blob/0a9753593d3e9f833fffe98100c33c0deef4c8a5/src/vs/workbench/browser/parts/notifications/notificationsToasts.ts#L186-L196)).
